cilium/CFP-28427-fqdn-selectors.md

```diff
@@ -0,0 +1,409 @@
+
+#### Background
+
+Cilium allocates security identities locally for certain types of destinations. Remote IPs discovered by ToFQDN policies are one source of local identities. In order to prevent traffic drops on restart, Cilium **attempts to maintain a stable mapping between IP and numeric identity between restarts**.
+
+The system as of v1.15 reads the "outgoing" BPF IPCache, which provides Prefix :: Numeric identity mappings. It then uses heuristics to reconstruct the likely set of labels for that prefix, and attempts to ensure the same prefix gets the same numeric identity. One all sources of metadata have synchronized, the IPCache proceeds with determining labels and generating the new BPF map.
+
+Since, in v1.15 and before, the only possible labels are the CIDR labels themselves and the `reserved:kube-apiserver` label, the simple 1:1 mapping between identity and prefix is simple to recreate.
+
+#### Proposal
+
+As always, the numeric identity for a given prefix should, as much as possible, be stable. The reconstruction heuristics will need to change, as a single ID can correspond to multiple prefixes. We can no longer use a simple heuristic to re-create identities.
+
+However, all hope is not lost. The state of the FQDN policy engine is also checkpointed to disk, so we can mostly recreate the set of labels that apply to a given IP. Additionally, we will **checkpoint the local allocator state to disk**, so that the same set of labels receives the same identity.
+
+Checkpointing the local allocator is required so that policy maps (which contain only numeric IDs) remain stable across restarts.
+
+The full procedure is as follows:
+
+1. Read the most recent allocator state. This maps Labels -> Numeric Identity (NID). Insert this in the allocator.
+2. Read the IPCache state. This maps Prefix -> Numeric Identity
+3. For any NIDs in the IPCache and not in the allocator, use the "withold + request" mechanism to request the same NID again
+4. Wait the synchronization grace period (10 minutes)
+5. Release all restored references (identity allocations, placeholder ipcache metadata)
+
+In this proposal, the Label -> NID mapping is now fixed. Thus, for prefixes to have stable NIDs across restarts, they must have the same labels. The agent waits for its data sources to synchronize before computing labels, so this should primarily be the case.
+
+#### Example
+
+The IPCache has three prefixes, each with a corresponding identity.
+
+- `192.168.0.0/24` -> `(cidr:192.168.0.0/24)`
+- `1.1.1.1/32` -> `(fqdn:*, fqdn:one.one.one.one)`
+- `2.2.2.2/32` -> `(fqdn:*)`
+
+If, on agent restart, all prefixes have the same set of labels, then the identity mapping is stable. If, however, the FQDN checkpointing is not exact (which does happen when an IP is near expiration), then `1.1.1.1/32` may change labels to `(fqdn:*)`. However, since the identity for `fqdn:*` is already allowed by policy maps, no traffic to `fqdn:*` is dropped.
```

Comment:
side comment: I suppose checkpointing is also not 100% foolproof either. IIRC the current name tracking per-endpoint is done every 5s or so, so there's always a race condition of whether the most up-to-date info is checkpointed to disk. Depending on the checkpointing mechanism there could be a similar blip with this as well.